### ISSUE TITLE
colorquantization, not averaging the last cycle

### ISSUE THREAD

**Issue by MiYanni** [jankovicsandras/imagetracerjava#2]
In colorquantization, you have the averaging set up so that it only occurs after the first cycle `(cnt>0)`, but it doesn't occur after the last cycle. Is this intended? Wouldn't it make sense to remove the `if(cnt>0)` check and put the averaging logic after the pixel looping in the cycles loop? Then, it would occur after every cycle. Also, it would mean that the accumulator (paletteacc) only needs to exist within the loop itself, since it gets created, set, and consumed in the same cycle.

Edit: What I was going to say is move the whole k-loop after the pixels loop; not split it into 2 loops. Then, simply not do the randomization part if it is the last cycle.

**Comment by jankovicsandras** [comment]
Hi,

It's partly intended, but you're right, that the averaging step can be moved. The point with `cnt>0` check is that the palette must not be changed if the user sets `options.colorquantcycles = 0`, because the user wants to use a fixed palette. (But the pixels must be indexed anyway.) So the averaging step should happen only after the first iteration.
The color randomizing step should not happen at the end of the last iteration, because it would set weird new colors to a few pixels.
It was probably a lazy, simple solution to keep them together, so the averaging step can be moved after the pixels loop (see below). However this may not be necessary, because if the missing last averaging step would lead to bad quality, just do another cycle. E.g. setting `options.colorquantcycles = 4` (instead of the default 3) would have very similar quality.

The following code is NOT tested (lack of time), it might be wrong:

``` java
        // Repeat clustering step "cycles" times
        for(int cnt=0;cnt<cycles;cnt++){

            // Randomize colors from the second iteration
            if(cnt>0){
                float ratio;
                for(int k=0;k<palette.length;k++){
                    /*// averaging TODO: moved below, test whether it works
                    if(paletteacc[k][3]>0){
                        palette[k][0] = (byte) (-128 + paletteacc[k][0] / paletteacc[k][4]);
                        palette[k][1] = (byte) (-128 + paletteacc[k][1] / paletteacc[k][4]);
                        palette[k][2] = (byte) (-128 + paletteacc[k][2] / paletteacc[k][4]);
                        palette[k][3] = (byte) (-128 + paletteacc[k][3] / paletteacc[k][4]);
                    }*/
                    ratio = (float)( (double)(paletteacc[k][4]) / (double)(imgd.width*imgd.height) );

                    // Randomizing a color, if there are too few pixels and there will be a new cycle
                    if((ratio<minratio)&&(cnt<(cycles-1))){
                        palette[k][0] = (byte) (-128+Math.floor(Math.random()*255));
                        palette[k][1] = (byte) (-128+Math.floor(Math.random()*255));
                        palette[k][2] = (byte) (-128+Math.floor(Math.random()*255));
                        palette[k][3] = (byte) (-128+Math.floor(Math.random()*255));
                    }

                }// End of palette loop
            }// End of Randomize colors from the second iteration

            // Reseting palette accumulator for averaging
            for(int i=0;i<palette.length;i++){
                paletteacc[i][0]=0;
                paletteacc[i][1]=0;
                paletteacc[i][2]=0;
                paletteacc[i][3]=0;
                paletteacc[i][4]=0;
            }

            // loop through all pixels
            for(int j=0;j<imgd.height;j++){
                for(int i=0;i<imgd.width;i++){

                    idx = ((j*imgd.width)+i)*4;

                    // find closest color from palette by measuring (rectilinear) color distance between this pixel and all palette colors
                    cdl = 256+256+256+256; ci=0;
                    for(int k=0;k<palette.length;k++){

                        // In my experience, https://en.wikipedia.org/wiki/Rectilinear_distance works better than https://en.wikipedia.org/wiki/Euclidean_distance
                        c1 = Math.abs(palette[k][0]-imgd.data[idx]);
                        c2 = Math.abs(palette[k][1]-imgd.data[idx+1]);
                        c3 = Math.abs(palette[k][2]-imgd.data[idx+2]);
                        c4 = Math.abs(palette[k][3]-imgd.data[idx+3]);
                        cd = c1+c2+c3+(c4*4); // weighted alpha seems to help images with transparency

                        // Remember this color if this is the closest yet
                        if(cd<cdl){ cdl = cd; ci = k; }

                    }// End of palette loop

                    // add to palettacc
                    paletteacc[ci][0] += 128+imgd.data[idx];
                    paletteacc[ci][1] += 128+imgd.data[idx+1];
                    paletteacc[ci][2] += 128+imgd.data[idx+2];
                    paletteacc[ci][3] += 128+imgd.data[idx+3];
                    paletteacc[ci][4]++;

                    arr[j+1][i+1] = ci;
                }// End of i loop
            }// End of j loop

            // Average colors from the second iteration TODO: test
            if(cnt>0){
                // averaging paletteacc for palette
                for(int k=0;k<palette.length;k++){
                    // averaging
                    if(paletteacc[k][3]>0){
                        palette[k][0] = (byte) (-128 + (paletteacc[k][0] / paletteacc[k][4]));
                        palette[k][1] = (byte) (-128 + (paletteacc[k][1] / paletteacc[k][4]));
                        palette[k][2] = (byte) (-128 + (paletteacc[k][2] / paletteacc[k][4]));
                        palette[k][3] = (byte) (-128 + (paletteacc[k][3] / paletteacc[k][4]));
                    }
                }
            }

        }// End of Repeat clustering step "cycles" times
```
